### ISSUE TITLE
Passing information to partials

### ISSUE THREAD

**Issue by w-biggs** [vapid/vapid#66]
**Is your feature request related to a problem? Please describe.**
Say I have a `{{> head}}` partial that includes everything inside my `<head>` tag. Inside the head is a `<title>` tag. If I want to set the title to something different in each page, I, well, can't. There doesn't seem to be any way to pass data to a partial, like you normally can with handlebars-style templating engines.

**Describe the solution you'd like**
The ability to pass custom parameters to a partial, like `{{> head title="About Us"}}`.

**Describe alternatives you've considered**
I can just not use partials for the `<head>` (or any other element that's mostly repeated with minor differences in each page, like a `<nav>`), but that makes the code very non-DRY and kinda takes away a lot of the advantage of using a templating engine.

**Additional context**
Most handlebars-style templating engines (ex. Panini) let you do this.

**Comment by srobbin** [comment]
@w-biggs Thanks, this is a good request.

How did you imagine we'd reference "title" in the partial? Like this?

```
<!DOCTYPE html>
<html>
<head>
  <title>{{title}}</title>
</head>
<body>
```

The issue I see here is that `{{title}}` would then be interpreted as a field in the General section. Maybe it could be written with an underscore, so it stayed a part of the private namespace (the parser currently ignores underscored variables, except `_permalink`, `_created_at`, and `_updated_at`).

```
<title>{{_title}}</title>
```

Or we could take a different approach, and instead of passing variables to a partial, we could use [front matter](https://github.com/jxson/front-matter), so that a page can have variables (which partials can also access). [Middleman uses this](https://middlemanapp.com/basics/frontmatter/), and it's pretty nice.

```
# index.html
---
title: My Page
---
{{> head}}
```

```
# _head.html
<title>{{_page.title}}</title>
```

**Comment by Jack89ita** [comment]
@srobbin Nice solution! I'm currently having the same problem, not for normal pages in which i include different partials, but in listing item page in which i do not have the control of the "current item". Infact i am not able to differentiate items title and description.
Any tips about this?
Thanks

**Comment by srobbin** [comment]
@Jack89ita I'm not sure I entirely understand, but it sounds like a good topic to bring up in the [Vapid forums](https://forums.vapid.com/).

**Comment by walidvb** [comment]
@srobbin would you accept PRs for this?

**Comment by srobbin** [comment]
@walidvb Yes, I'm open to a PR, but I think we should probably discuss the implementation a little bit first. Were you thinking of using [front matter](https://github.com/vapid/vapid/issues/66#issuecomment-427440325)?

**Comment by walidvb** [comment]
@srobbin i might have spoken too fast, was speaking with another friend about this and we were both needing the functionality to improve DRY(as per #47 ).

I have a rails and haml background, so i would go with `{> partial variable=value}`. My opinion is that partials should be able to receive variables, and the frontmatter solution actually brings global variables to a template/html file. 
This could maybe be solved by adding a `{{variable hidden=true default=default_value}}`?

**Comment by srobbin** [comment]
@walidvb Re: frontmatter and partials, I was thinking it would work something like this:

```
# index.html
---
title: My Page
---
{{> head }}

{{> another foo=bar}}
```

```
# _head.html
<title>{{_page.title}}</title>
```

```
# _another.html
<h1>{{_page.title}}</h1>
<p>{{_page.foo}}</p>
```

Where partial variables are merged into the frontmatter "_page" namespace. That might provide a nice way to set values for both the template and partial.

Also, re: DRY, this might naturally lead to a discussion of layouts.

**Comment by walidvb** [comment]
Can this be done in 2 steps?
First, accept variables from the the direct syntax `{ > partial varName=varValue}`, then add front-matter for global variables? what do you think?

**Comment by walidvb** [comment]
do we really want to close stale issues? :/

**Comment by enigmatic-user** [comment]
Starting my first Vapid project, I also miss this feature very much. I hope the discussion here hasn't died...

**Comment by webuxr** [comment]
@srobbin Just read through the thread above and noticed [your comment](https://github.com/vapid/vapid/issues/66#issuecomment-437980172) above. Interesting that this comment was almost a year ago! Also interesting that you and I just chatted about this in the [Vapid Forum recently](https://forums.vapid.com/t/loading-partials-with-tabular-data-from-existing-json-files/327/2?u=webuxr).

In reference to using Front Matter, I seem to remember reading somewhere about the topic and 1 requirement for it to work (i.e.: storing simple key-value pairs within the same document as say, a page's Markdown content) is that Front Matter needs to start with  `---` as the first line of the file. The other requirement, I believe is, that documents containing both Front Matter and content (again, this is usually Markdown) needed to be stored in a separate directory (say `/www/pages/`) than partials (`/www/partials/`) or layouts or templates (something that Vapid does not currently have).

If Front Matter was stored in a `/www/pages/` directory, then there would be several benefits we could take advantage of during the build process. Things like pre-compilation, parsing YAML and Markdown, etc..

Thoughts about resurfacing the discussion about Front Matter?

**Comment by webuxr** [comment]
EDIT: I just remembered where I saw this info about Front Matter. It was from [Zurb's Foundation for Emails project](https://foundation.zurb.com/emails/docs/sass-guide.html) where they say this:

> ...breakdown of the files in the `/src` folder:
>
> `assets/`: Sass and image files.
> `layouts/`: Boilerplate HTML that wraps all of your emails.
> `pages/`: HTML files for emails.
> `partials/`: Reusable chunks of HTML that can be injected into pages.

Good stuff there!

Obviously Vapid does not need to inline styles or use a special email-friendly HTML variation (aka: Zurb's Inky language), but their project's structure and Gulp-based build system might have some good ideas that could be considered for Vapid. Maybe...

**Comment by amiller-gh** [comment]
So, there was a [pretty big conversation](https://github.com/emberjs/rfcs/pull/449) about this over in Ember land throughout the Glimmer (next gen Handlebars) buildout process. Here is the syntax they settled on for parameters passed to components to enable template-only component composition: https://glimmerjs.com/guides/templates-and-helpers 

A similar sigil can be used for referencing passed attributes to partials.

On a very different note, in our new component focused world, partials are often more trouble than they're worth for exactly the reasons you describe above – you can never be quite sure what is being passed to them and in what context they are being rendered with. (At least Hbs doesn't have the problems Dustjs had – partials [*_automatically inherited their parent's context_*](https://www.dustjs.com/guides/partials/) 😱)

The industry has slowly begun to consolidate around components instead of partials because they have that very clean boundary between render contexts – with params and events being the only interface between them ([DDAU](https://medium.com/@peter.yun.kim/react-data-down-action-up-6595ffbe6be1) for the win).

So, that being said, Vapid has an interesting situation – it renders to a static site, so there are no actions, and context inheritance has *never* worked. If argument passing works, that means that functionally partials *are* static site components in vapid.

IMO I think it would behoove Vapid to move toward similar component constructs like frontend frameworks and remove partials entirely in favor of a more component-like syntax. So, instead of

```hbs
<h1>Awesome Website</h1>
{{> my-partial value=whatever}}
```

You would write:
```hbs
<h1>Awesome Website</h1>
<MyPartial @value=whatever />
```

Unlike partials, the lookup path of a component is not specified in its invocation, so Vapid would have to make some opinions about where components live. As @webuxr pointed out above, many platforms choose to do this with a dedicated directory. Many other frameworks have chosen to put components in the `/components` directory of a project structure – Vapid can do the same. This way, the `<MyPartial>` component above would live in `components/MyPartial.html`

Moving to component syntax and dropping partials has a number of benefits:

 - Industry Consolidation: We're not bucking the trend here, inventing new syntax, or using features that are losing support. The syntax will look familiar to newcomers.
 - Static analysis of component files can surface required parameters that *must* be passed to components at their invocation site, and throw a helpful error when they are not passed.
 - The global `section` states can remain global in these components because of the `@` sigil. 
 - Having an opinionated location for components will clean up the regular mess of partials that end up floating around handlebars projects.
 - Clears up confusion between utility partials and individual repeating content partials – an interesting bit of syntax collision currently lurking in Vapid.

**Comment by amiller-gh** [comment]
Oh, also, ref #157. Just like I mentioned in #57, custom components could be delivered in a `/www/components`

A vapid plugin could provide custom directives in a `/www/components` folder. They would be automatically discovered, merged into the parent project (giving priority to locally defined directives of the same name `Object.assign` style), and made available in the parent project's templates.

**Comment by amiller-gh** [comment]
Ooh, and! If we want to totally be rid of the concept of the partials, we can add a new option for rendering individual items in content collections by adding a `/www/collections` directory in liu of the magic `_sectionName.html` partials. 

In `/www/collections` you would provide html files whos name match the name of a repeating section. Because of the naming convention, we can automatically wrap these in the section scope, removing the need for the mandatory `{{section whatever}}` on the magical section partials.

Also, all of the above changes can be implemented as new features alongside the existing partial implementation, maintaining backward compatibility. Encountered partials can log a deprecation warning and point to migration documentation. (A migration script wouldn't be hard to put together either if we want to go that route.)
